src/c/quikr_train.1
.TH quikr_train 1 quikr_train-2013-09
.SH NAME
quikr_train \- train databases for use with Quikr
.SH SYNOPSIS
.B quikr_train
.RB [ \-i
.IR input]
.RB [ \-o
.IR output]
.RB [ \-k
.IR kmer ]
.RB [ \-v ]
.P
.BR quikr " ..."
.SH DESCRIPTION
.B quikr
The quikr_train is a utility to train a database for use with quikr.
Before running the quikr utility, you need to generate the sensing matrix or
download a pretrained one from our project's homepage.

quikr_train returns a sensing matrix that can be used with the quikr
function. 
.P
.SH OPTIONS
.TP
.B \-i, --input
the database of sequences to create the sensing matrix. (fasta format)
.TP
.B \-k, --kmer
specify what size of kmer to use. (default value is 6)
.TP
.B \-o, --output
the sensing matrix. (a gzip'd text file)
.TP
.B \-v, --verbose
verbose mode.
.TP
.B \-V, --version
print version.
.SH EXAMPLES
Use quikr_train to generate a sensing matrix from rdp7.fasta. This uses 6mers by default.
.P
quikr_train -i rdp7.fa -o rd7_sensing_matrix.gz
.SH USAGE
If you do not have a .gz file on your output matrix name, it will be appended.
.SH "SEE ALSO"
\fBmultifasta_to_otu\fP(1), \fBquikr\fP(1).
.SH AUTHORS
.B quikr
was written by Gail Rosen <gailr@ece.drexel.edu>, Calvin Morrison 
<mutantturkey@gmail.com>, David Koslicki, Simon Foucart, and Jean-Luc Bouchot.
.SH REPORTING BUGS
.TP
Please report all bugs to Gail Rosen <gailr@ece.drexel.edu>. Include your \
operating system, current compiler, and test files to reproduce your issue.
.SH COPYRIGHT.
Copyright \(co 2013 by Calvin Morrison and Gail Rosen.  Permission to use, 
copy, modify, distribute, and sell this software and its documentation for
any purpose is hereby granted without fee, provided that the above copyright 
notice appear in all copies and that both that copyright notice and this 
permission noticeappear in supporting documentation.  No representations are
made about the suitability of this software for any purpose.  It is provided
"as is" without express or implied warranty.



CHANGELOG
May 15th 2013, 1.0.0 Released
- initial version of quikr C version
- initial version of quikr python version
- initial version of quikr matlab scripts

May 16th 2013, 1.0.1 Released
- remove extra data 
- remove old multifasta_to_otu matlab scripts
- minor documentation changes

May 20th 2013, 1.0.2 Released
- correted usage strings
- install manpages globally (to {PREFIX}/share/man/man1)

June 12th 2013, 1.0.3 Relased
- licensing files and documentation added
- removed unused code from the NBC project
- fixed a overflow bug in quikr_train
- added native gzip decoding for quikr and multifasta_to_otu
- documentation fixes
- use int64_t for NNLS to avoid overflow issues
- use long for multifasta/quikr to avoid overflow issues
- implement --version and -V in quikr
- fix input directory check in multifasta
- matlab: don't depend on mer text files, instead use generate_kmers for kmers

Sept 5th 2013, 1.0.4 Released
- fix -h and --help options to parse correctly
- remove the debugging option
- update Makefile to ensure NBC and the C verison are build before installing
- documentation updates


src/c/quikr_train.c
#include <ctype.h>
#include <errno.h>
#include <getopt.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <zlib.h>

#include "quikr_functions.h"

#define USAGE "Usage:\n\tquikr_train [OPTION...] - train a database for use with quikr.\n\nOptions:\n\n-i, --input\n\tthe database of sequences to create the sensing matrix (fasta format)\n\n-k, --kmer\n\tspecify what size of kmer to use. (default value is 6)\n\n-o, --output\n\tthe sensing matrix. (a gzip'd text file)\n\n-v, --verbose\n\tverbose mode.\n\n-V, --version\n\tprint version."

int main(int argc, char **argv) {

  char probabilities_command[512];
  char *line = NULL;
  char *val;
  size_t len = 0;


  int c;
  int kmer = 6;

  char *fasta_file = NULL;
  char *output_file = NULL;

  int x = 0;
  int y = 0;

  int verbose = 0;

  gzFile output = NULL;

  while (1) {
    static struct option long_options[] = {
      {"verbose", no_argument, 0, 'v'},
      {"help", no_argument, 0, 'h'},
      {"version", no_argument, 0, 'V'},
      {"input", required_argument, 0, 'i'},
      {"kmer",  required_argument, 0, 'k'},
      {"output", required_argument, 0, 'o'},
      {0, 0, 0, 0}
    };

    int option_index = 0;

    c = getopt_long (argc, argv, "i:o:k:hvV", long_options, &option_index);

    if (c == -1)
      break;

    switch (c) {
      case 'i':
        fasta_file = optarg;
        break;
      case 'k':
        kmer = atoi(optarg);
        break;
      case 'o':
        output_file = optarg;
        break;
      case 'v':
        verbose = 1;
        break;
      case 'V':
        printf("%s\n", VERSION);
        exit(EXIT_SUCCESS);
        break;
      case 'h':
        printf("%s\n", USAGE);
        exit(EXIT_SUCCESS);
        break;
      case '?':
        /* getopt_long already printed an error message. */
        break;
      default:
        exit(EXIT_FAILURE);
    }
  }


  if(fasta_file == NULL) {
    fprintf(stderr, "Error: input fasta file (-i) must be specified\n\n");
    fprintf(stderr, "%s\n", USAGE);
    exit(EXIT_FAILURE);
  }

  if(output_file == NULL) {
    fprintf(stderr, "Error: output matrix file (-o) must be specified\n\n");
    fprintf(stderr, "%s\n", USAGE);
    exit(EXIT_FAILURE);
  }

  if(verbose) {
    printf("kmer size: %d\n", kmer);
    printf("fasta file: %s\n", fasta_file);
    printf("output file: %s\n", output_file);
  }

  if(access (fasta_file, F_OK) == -1) {
    fprintf(stderr, "Error: could not find %s\n", fasta_file);
    exit(EXIT_FAILURE);
  }

  if(strcmp(&output_file[strlen(output_file) - 3], ".gz") != 0) {
    char *temp = malloc(strlen(output_file) + 4);
    if(temp == NULL) {
      fprintf(stderr, "Could not allocate enough memory\n"); 
      exit(EXIT_FAILURE);
    }
    sprintf(temp, "%s.gz", output_file);
    output_file = temp;
    printf("appending a .gz to our output file: %s\n", output_file);
  }

  // 4 ^ Kmer gives us the width, or the number of permutations of ACTG with kmer length
  int width = pow(4, kmer);
  int sequences = count_sequences(fasta_file);
  if(sequences == 0) {
    fprintf(stderr, "Error: %s contains 0 fasta sequences\n", fasta_file);
  }

  if(verbose)
    printf("sequences: %d\nwidth: %d\n", sequences, width);

  // Allocate our matrix with the appropriate size, just one row
  double *trained_matrix = malloc(width*sizeof(double));
  if(trained_matrix == NULL) {
    fprintf(stderr, "Could not allocate enough memory\n");
    exit(EXIT_FAILURE);
  }

  // call the probabilities-by-read command
  sprintf(probabilities_command, "generate_kmers %d | probabilities-by-read %d %s /dev/stdin", kmer, kmer, fasta_file);
  FILE *probabilities_output = popen(probabilities_command, "r");
  if(probabilities_output == NULL) {
    fprintf(stderr, "Error could not execute: %s\n", probabilities_command);
    exit(EXIT_FAILURE);
  }

  // open our output file
  output = gzopen(output_file, "w6");
  if(output == NULL) {
    fprintf(stderr, "Error: could not open output file, error code: %d", errno);
    exit(EXIT_FAILURE);
  }

  if(verbose)
    printf("Writing our sensing matrix to %s\n", output_file);

  // read normalize and write our matrix  in one go
  for(x = 0; x < sequences; x++) {

    getline(&line, &len, probabilities_output);

    // Read our first element in outside of the loop
    val = strtok(line,"\t\n\r");
    trained_matrix[0] = atof(val);
    // iterate through and load the array
    for (y = 1; y < width; y++) {
      val = strtok (NULL, "\t\n\r");
      trained_matrix[y] = atof(val);
    }

    double row_sum = 0;

    for( y = 0; y < width; y++) {
      row_sum = row_sum + trained_matrix[y];
    }

    for( y = 0; y < width; y++) {
      trained_matrix[y] = trained_matrix[y] / row_sum;
    }

    for( y = 0; y < (width - 1); y++) {
      gzprintf(output, "%.10f\t", trained_matrix[y]);
    }
    gzprintf(output, "%.10f\n", trained_matrix[width]);
  }

  free(trained_matrix);
  gzclose(output);
  pclose(probabilities_output);

  return 0;
}
